lib/dRep/dRepImpl.py

```diff
@@ -0,0 +1,441 @@
+                bc = BinnedContigs(upa, get_bins_dir='download')
+                
+                if bc.is_empty():
+                    # never mention this bc again
+                    params['genomes_refs'].remove(upa)
+                    BinnedContigs.created_instances.remove(bc)
+                
+
+            if len(params['genomes_refs']) == 0:
+                msg = 'Sorry, please input at least one non-empty BinnedContigs'
+                raise ValueError(msg)
+
+
+
+        #
+        ##
+        ###
+        ####
+        #####
+
+        for bc in BinnedContigs.created_instances:
+            bc.calc_stats()
+
+
+
+        #
+        ##
+        ### pool
+        ####
+        #####
+        
+ 
+        binsPooled_dir = os.path.join(_globals.run_dir, 'binsPooled')
+        os.mkdir(binsPooled_dir)
+
+        for bc in BinnedContigs.created_instances:
+            bc.pool_into(binsPooled_dir)
+        
+        dprint("os.listdir(binsPooled_dir)", run={**locals(), **globals()})
+
+
+        #
+        ##
+        ### 
+        #### params
+        #####
+       
+
+        # flatten param groups (if any)
+
+        paramGrp_key_l = ['filtering', 'genome_comparison', 'clustering', 'scoring', 'warnings']
+
+        for paramGrp_key in paramGrp_key_l:
+            if paramGrp_key in params:
+                paramGrp_d = params[paramGrp_key]
+                for paramIndiv_key in paramGrp_d:
+                    params[paramIndiv_key] = paramGrp_d[paramIndiv_key]
+                params.pop(paramGrp_key)
+
+
+        #
+
+        dRep_params = ['--debug']
+
+
+        #
+        
+        if _globals.debug and 'processors' in params:
+            dRep_params.extend(['--processors', params.get('processors')])
+        else:
+            dRep_params.extend(['--processors', '8'])
+
+                
+        # extract non-default parameters
+
+        dRep_param_defaults = {
+                'length': 50000,
+                'completeness': 75,
+                'contamination': 25, 
+                'ignoreGenomeQuality': "False", 
+                'MASH_sketch': 1000,
+                'S_algorithm': 'ANImf',
+                'n_PRESET': 'normal',
+                'P_ani': 0.9,
+                'S_ani': 0.99,
+                'SkipMash': "False", 
+                'SkipSecondary': "False",
+                'cov_thresh': 0.1,
+                'coverage_method': 'larger',
+                'clusterAlg': 'average', 
+                'completeness_weight': 1,
+                'contamination_weight': 5,
+                'strain_heterogeneity_weight': 1, 
+                'N50_weight': 0.5,
+                'size_weight': 0,
+                'warn_dist': 0.25, 
+                'warn_sim': 0.98,
+                'warn_aln': 0.25,
+                'checkM_method': 'lineage_wf',
+                }
+
+        params_bool = [key for key in dRep_param_defaults if dRep_param_defaults[key] == 'False']
+
+        for key in dRep_param_defaults:
```

Comment:
I'm having a hard time understanding the purpose of this section. Maybe it could use a couple more comments 
Why not just do something like:
```py
dRep_params = []
for (key, val) in user_params:
    dRep_params.append(f"--{key}")
    if val is not None:
        dRep_params.append(str(val))
```

If "False" is supposed to indicate no value, I would instead use None or empty string and filter those out.

What's the purpose of checking against defaults? Why not just pass the user params to dRep?

Reply:
a lot of this bad code may be due to the fact i had/have no idea how the narrative ui works.

the ui lets you choose "True" or "False", and to the program dRep, these are kind of binary and set by the presence of `--parameter`, not `--parameter bool`

i'll try to comment/fix this or figure out how the narrative works

Reply:
i check against defaults because the ui has defaults, and those would get passed in, and it looks messy if you just want to do a simple run but all the defaults are explicitly passed in. i return the command in a report